cadc-dali/src/main/java/ca/nrc/cadc/dali/tables/parquet/ParquetWriter.java

```diff
@@ -0,0 +1,205 @@
+        this.formatFactory = ff;
+    }
+
+    @Override
+    public void write(VOTableDocument voTableDocument, OutputStream out) throws IOException {
+        write(voTableDocument, out, Long.MAX_VALUE);
+    }
+
+    @Override
+    public void write(VOTableDocument voTableDocument, OutputStream out, Long maxRec) throws IOException {
+        log.debug("ParquetWriter Write service called. MaxRec = " + maxRec);
+        for (VOTableResource resource : voTableDocument.getResources()) {
+            Schema schema = DynamicSchemaGenerator.generateSchema(resource.getTable().getFields());
+            OutputFile outputFile = outputFileFromStream(out);
+
+            TableData tableData = resource.getTable().getTableData();
+            resource.getTable().setTableData(null);
+
+            StringWriter stringWriter = new StringWriter();
+            VOTableWriter votableWriter = new VOTableWriter();
+            votableWriter.write(voTableDocument, stringWriter, maxRec);
+
+            Map<String, String> customMetaData = new HashMap<>();
+            customMetaData.put("votable", stringWriter.toString());
+
+            try (org.apache.parquet.hadoop.ParquetWriter<GenericRecord> writer = AvroParquetWriter.<GenericRecord>builder(outputFile)
+                    .withSchema(schema)
+                    .withCompressionCodec(CompressionCodecName.SNAPPY)
+                    .withRowGroupSize(Long.valueOf(org.apache.parquet.hadoop.ParquetWriter.DEFAULT_BLOCK_SIZE))
+                    .withPageSize(org.apache.parquet.hadoop.ParquetWriter.DEFAULT_PAGE_SIZE)
+                    .withConf(new Configuration())
+                    .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
+                    .withValidation(false)
+                    .withDictionaryEncoding(false)
+                    .withExtraMetaData(customMetaData)
+                    .build()) {
+
+                Iterator<List<Object>> iterator = tableData.iterator();
+                int recordCount = 1;
+
+                while (iterator.hasNext() && recordCount <= maxRec) {
+                    GenericRecord record = new GenericData.Record(schema);
+                    List<Object> rowData = iterator.next();
+                    int columnIndex = 0;
+
+                    for (Schema.Field field : schema.getFields()) {
+                        String columnName = field.name();
+                        record.put(columnName, rowData.get(columnIndex)); // TODO: convert non-simple row data to correct format before adding to record
+                        columnIndex++;
+                    }
+
+                    writer.write(record);
+                    recordCount++;
+                    log.debug("Total Records generated= " + (recordCount - 1));
+                }
+            } catch (Exception e) {
+                throw new IOException("error while writing", e);
+            }
+        }
+        out.close();
+    }
+
+    @Override
```

Comment:
I think we should remove the two TableWriter.write(..., Writer, ...) methods from the interface. It's clearly a bad idea since it only applies to text-based formats. Then these dummy methods could be removed from here.

It would be OK to keep those methods as convenience in the inherently text (AsciiTableWriter and VOTableWriter) but would need to remove the `@Override`. 

---

I am less sure what we should do about `write(Throwable, OutputStream)` here... I guess since getErrorContentType() says "text/plain" it should output a plain text error message (for now at least). There are reasons, but complicated and from othewr IVOA standards.